src/axolotl/utils/callbacks/qat.py

```diff
@@ -0,0 +1,50 @@
+"""
+QAT Callback for HF Causal Trainer
+"""
+
+import logging
+from functools import partial
+
+from torch import nn
+from torchao.quantization.qat.embedding import FakeQuantizedEmbedding
+from torchao.quantization.qat.linear import FakeQuantizedLinear
+from transformers import TrainerCallback
+
+from axolotl.utils.schemas.quantization import QATConfig
+
+LOG = logging.getLogger(__name__)
+
+
+def toggle_fake_quant(mod: nn.Module, enable: bool):
+    """
+    Toggle fake quantization for any fake quantized linear or embedding layers in the model.
+
+    Args:
+        mod: The module to toggle fake quantization for.
+        enable: Whether to enable or disable fake quantization.
+    """
+    if isinstance(mod, (FakeQuantizedLinear, FakeQuantizedEmbedding)):
+        if (
+            isinstance(mod, FakeQuantizedLinear)
+            and mod.activation_fake_quantizer is not None
+        ):
+            mod.activation_fake_quantizer.enabled = enable
+        mod.weight_fake_quantizer.enabled = enable
+
+
+class QATCallback(TrainerCallback):
+    """
+    Callback to toggle fake quantization for the model.
+    """
+
+    def __init__(self, cfg: QATConfig):
+        self.cfg = cfg
+
+    def on_step_begin(self, args, state, control, model, **kwargs):
+        if self.cfg.fake_quant_after_n_steps is not None:
+            if state.global_step == 0:
+                LOG.info(f"Disabling fake quantization at step {state.global_step}")
+                model.apply(partial(toggle_fake_quant, enable=False))
```

Comment:
stupid question: can we avoid this by toggling at the start, rather than checking for this on the first step of training?